### DISCUSSION TITLE
Section 8.4 Behavior Module

### DISCUSSION THREAD

**Discussion by kczimm** [danieleades/rtps#8] [Ideas]
I'm reading through [Section 8.4](https://www.omg.org/spec/DDSI-RTPS/2.5/PDF#page=72) of the specification. The behavior module looks like it will practically end up being implemented in the structure module `impl` blocks. We will need an implementation of the `HistoryCache` for the `Reader` and the `Writer`. I'm not sure where this should be placed within the module hierarchy. 

Perhaps this discussion can be a place for brainstorming the approach to the behavior module. The message and the structure module more or less map to a Rust model, but I believe the behavior model will not be the same approach. Any thoughts would be appreciated.

**Comment by danieleades** [comment]
> I'm reading through [Section 8.4](https://www.omg.org/spec/DDSI-RTPS/2.5/PDF#page=72) of the specification. The behavior module looks like it will practically end up being implemented in the structure module `impl` blocks. We will need an implementation of the `HistoryCache` for the `Reader` and the `Writer`. I'm not sure where this should be placed within the module hierarchy.
> 
> Perhaps this discussion can be a place for brainstorming the approach to the behavior module. The message and the structure module more or less map to a Rust model, but I believe the behavior model will not be the same approach. Any thoughts would be appreciated.

hi @kczimm. Apologies, it's been a crazy week of work, and i've not had any opportunity to look at this. I like where your head's at, and i've been thinking along the exact same lines. I have some ideas for how the structure and behaviour modules can be separated, hopefully i get an opportunity to play with this sometime in the next week or so

**Reply by danieleades** [reply]
@kczimm i've started playing with this in a branch - https://github.com/danieleades/rtps/tree/pim-crate

I'm not yet happy with the structure of the exports, and in particular how they relate to the generated documentation, so expect a lot of churn.

It's a layered approach. the general idea is to place all of the platform-independent types into a separate crate. Platform-specific implementations (including the UDP implementation) will then depend on and extend the PIM crate.

That allow down stream implementations to

- share PIM code
- be interoperable with each other

**Comment by danieleades** [comment]
I'm considering layering the UDP implementation as well, by excluding any concrete implementation of the publishers and subscribers.

the crate structure would then be

PIM -> PSM (UDP) -> UDP (implementation)

- PIM: provides generic types, traits, and interfaces.
- PSM (UDP): specialises the types and traits for the UDP model. Provides serialisation and deserialisation to common data representation (CDR).
- UDP (implementation): provides the concrete implementation of the publishers and subscribers.

layering in this way allows this crate to be an unopinionated foundational crate of other implementations. For example in a perfect universe, `RustDDS` would be based on this crate, but provide a concrete implementation depending on what QOS is important to their application.

@kczimm what do you think?

If we can agree on the structure, then i think the churn will be lessened, and I promise i can start merging your pull requests ;)

i've got a draft here - https://github.com/danieleades/rtps/pull/12

**Reply by kczimm** [reply]
Sorry for the long delay. Various things have kept me from this. But it all sounds good and I look forward to helping this move. I'll take a look at the draft you referenced.
